{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "dc0b9e3c-7e09-409d-bc4e-31841cc3a44c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-08-17 23:28:43.190958: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory\n",
      "2021-08-17 23:28:43.191038: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n"
     ]
    }
   ],
   "source": [
    "import requests as r\n",
    "import pandas as pd\n",
    "import json\n",
    "import numpy as np\n",
    "from tensorflow import keras\n",
    "from keras.preprocessing.sequence import pad_sequences\n",
    "import tensorflow as tf\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76b386e9-4790-486b-8fbd-01657407ed43",
   "metadata": {},
   "source": [
    "## Obtengo datasets\n",
    "\n",
    "Utilizando la API de [mobidb](https://mobidb.bio.unipd.it/) descargo en formato JSON los proteomas de _Homo sapiens_ , _Danio rerio_ , y _Gallus gallus_. Hago un backup de los datos en la carpeta `src/pickle/raw_mobidb_data` para no redescargar los datos.\n",
    "\n",
    "Luego examiné la longitud de secuencias para tratar de establecer cual sería la secuencia mas corta que resultase significativa, teniendo en cuenta que voy a tener que _paddear_ los datos para que todos los datos sean igualemente largos. Los datos con secuencias 'recortadas' quedan guardados en la variable `curated_data`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6164e9b7-46eb-4053-8502-38b84dc0cc83",
   "metadata": {},
   "outputs": [],
   "source": [
    "url='https://mobidb.bio.unipd.it/api/download?proteome=UP000005640&format=json'\n",
    "human_proteome_data=r.get(url).json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d86bf6e4-61ab-4339-bbc9-3cd29c83b158",
   "metadata": {},
   "outputs": [],
   "source": [
    "url_drosophila='https://mobidb.bio.unipd.it/api/download?proteome=UP000000803&format=json'\n",
    "drosophila_proteome_data=r.get(url_drosophila).json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "8c6a560b-b91d-4ba5-8bf3-ec2b2526d956",
   "metadata": {},
   "outputs": [],
   "source": [
    "url_celegans='https://mobidb.bio.unipd.it/api/download?proteome=UP000001940&format=json'\n",
    "celegans_proteome_data=r.get(url_celegans).json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "13856710-76ca-49a0-b950-53418a1b696a",
   "metadata": {},
   "outputs": [],
   "source": [
    "url_danio='https://mobidb.bio.unipd.it/api/download?proteome=UP000000437&format=json'\n",
    "danio_proteome_data=r.get(url_danio).json()\n",
    "\n",
    "url_gallus='https://mobidb.bio.unipd.it/api/download?proteome=UP000000539&format=json'\n",
    "gallus_proteome_data=r.get(url_gallus).json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "69368330-e06a-4456-a5bd-503fe69ec2d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "proteome_data=human_proteome_data+drosophila_proteome_data+gallus_proteome_data+danio_proteome_data+celegans_proteome_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "03d5d80a-e9e0-4080-a9a2-d231b6032d2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('pickle/raw_mobidb_data/proteomes.json', 'w') as file:\n",
    "    json.dump(proteome_data, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "6c04b4d7-f995-436e-b665-45c33a95e986",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences=[protein['sequence'] for protein in proteome_data]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "a7ac7b1a-ca3e-4afa-91d2-323bbb52940d",
   "metadata": {},
   "outputs": [],
   "source": [
    "seqlengths=[len(sequence) for sequence in sequences]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "7eb83c56-d22e-4e5f-9fa3-ac36404207af",
   "metadata": {},
   "outputs": [],
   "source": [
    "sl=sorted(seqlengths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "0800068a-09dd-4d3a-a7bf-3cf56c094288",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZEAAAD8CAYAAAC2PJlnAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAAdnElEQVR4nO3dfZBddZ3n8ffndnhwBUmCbCqVQIFrCsdxV8xElwBlOaRcAV3DTrmkE3dIsWij4pQuuyqsVbvOH1ursw8qW1a6W3EMOzyKUmRc1hkmMjOkw4NBEVF0aVGKpICsCHGQGtzu/u4f93dun759u/v26Xv6ntv386rquuf8zsP99iGcb/8ezu8oIjAzMyui1u0AzMysdzmJmJlZYU4iZmZWmJOImZkV5iRiZmaFOYmYmVlhpSURSWdLeiT382tJH5e0VtI9kp5In2vS/pJ0vaRxSY9K2lxWbGZm1hmlJZGI+GlEnBMR5wC/B7wM3AlcC+yPiE3A/rQOcDGwKf0MAXvKis3MzDpjuZqztgE/i4ingO3A3lS+F7g0LW8Hboy6B4DVktYvU3xmZlbAqmX6nkHglrS8LiKeScvPAuvS8gbg6dwxh1PZM7kyJA1Rr6kwcPyJv/eadWe0H4VU/yzylH6rYyXOeu2r+fkvf8PrTjtp8ec0M+uChx9++JcRcVonzlV6EpF0PPBe4LrmbRERkhZ1R4+IUWAUYO0ZZ8e2T462fWxtYBURU8TU1GK+cs5jVatx89BWdn35fm696nxqNY9TMLPqk/RUp861HHe9i4HvRcRzaf25rJkqfR5N5UeA03PHbUxlZmZWUcuRRHYy3ZQFsA/YnZZ3A3flyi9Po7TOBY7lmr3MzKyCSm3OkvRq4J3AVbnizwK3S7oSeAq4LJXfDVwCjFMfyXVFmbGZmdnSlZpEIuI3wKlNZc9TH63VvG8AV5cZj5mZdZZ7gs3MrDAnETMzK8xJxMzMCnMSMTOzwpxEzMysMCcRMzMrzEnEzMwKcxIxM7PCnETMzKwwJxEzMyvMSWQJIoKpAtPKm5mtFE4iSxHBrpGDUOAdV2ZmK4GTyBIpe+OhmVkfchIxM7PCnETMzKwwJxEzMyvMScTMzApzEjEzs8KcRMzMrDAnETMzK8xJxMzMCnMSMTOzwkpNIpJWS7pD0k8kPS5pq6S1ku6R9ET6XJP2laTrJY1LelTS5jJjMzOzpSu7JvJF4NsR8QbgzcDjwLXA/ojYBOxP6wAXA5vSzxCwp+TYzMxsiUpLIpJOAd4O3AAQEb+NiBeB7cDetNte4NK0vB24MeoeAFZLWl9WfJ2UzeYb4ZkYzay/lFkTOQv4v8CfSvq+pK9IejWwLiKeSfs8C6xLyxuAp3PHH05lM0gaknRI0qFXXjpWYviLEMHOkYNOImbWd8pMIquAzcCeiHgL8Bumm64AiPpdd1F33ogYjYgtEbHlhJNO6ViwS+XZfM2sH5WZRA4DhyPiwbR+B/Wk8lzWTJU+j6btR4DTc8dvTGVmZlZRpSWRiHgWeFrS2aloG/BjYB+wO5XtBu5Ky/uAy9MorXOBY7lmLzMzq6BVJZ//j4CbJB0PPAlcQT1x3S7pSuAp4LK0793AJcA48HLa18zMKqzUJBIRjwBbWmza1mLfAK4uMx4zM+ssP7FuZmaFOYmYmVlhTiIl8MOHZtYvnERKEBHsGD7gJGJmK17Zo7P6lqRGjSRb9wOJZrbSOIl0SJYw8okiItg5ehCAW68630nEzFYcN2d1yhzzZ0k1JF9mM1uZfHfrINc0zKzfOImYmVlhTiId5KG9ZtZvnEQ6ye8VMbM+4yTSYe4XMbN+4iRiZmaFOYl0WP4BQzOzlc5JpNMi2DVycJEv/TUz601OIiVwv4iZ9QsnkWXi4b9mthI5iSwTz+xrZiuRk8gycjOXma00TiJmZlaYk4iZmRXmJGJmZoWVmkQk/ULSDyU9IulQKlsr6R5JT6TPNalckq6XNC7pUUmby4zNzMyWbjlqIr8fEedExJa0fi2wPyI2AfvTOsDFwKb0MwTsWYbYzMxsCbrRnLUd2JuW9wKX5spvjLoHgNWS1nchvo7w9Cdm1g/KTiIB/KWkhyUNpbJ1EfFMWn4WWJeWNwBP5449nMpmkDQk6ZCkQ6+8dKysuJfO05+YWR9YVfL5L4iII5L+IXCPpJ/kN0ZESFrUbTYiRoFRgLVnnF3pW7SfCzGzla7UmkhEHEmfR4E7gbcBz2XNVOnzaNr9CHB67vCNqczMzCqqtCQi6dWSTs6WgX8GPAbsA3an3XYDd6XlfcDlaZTWucCxXLPXihERTE5OMjk56SlQzKznldmctQ64MzXprAJujohvS/oucLukK4GngMvS/ncDlwDjwMvAFSXGtqzynewRweDwAaQat334Ajd5mVlPKy2JRMSTwJtblD8PbGtRHsDVZcXTVamTvTYwANT7Spw8zGwl8BPry8RJw8xWIicRMzMrzEnEzMwKcxIxM7PCnETMzKwwJxEzMyvMScTMzApzEjEzs8KcRMzMrDAnETMzK8xJxMzMCnMSMTOzwpxEzMysMCcRMzMrzEnEzMwKcxIxM7PCnETMzKwwJxEzMyvMScTMzApzEumiiGBqaor66+XNzHqPk0gXRQQ7hg84iZhZz3ISWUZZzSNPUpeiMTNbutKTiKQBSd+X9K20fpakByWNS7pN0vGp/IS0Pp62n1l2bMsugl0jB8EVDzNbIZajJvIx4PHc+ueAz0fE64EXgCtT+ZXAC6n882m/Fcc1DzNbSUpNIpI2Au8GvpLWBVwI3JF22Qtcmpa3p3XS9m3qkzuuO9jNrFeVXRP5AvBJIOsIOBV4MSIm0vphYENa3gA8DZC2H0v7r0j5/hF3sJtZryotiUh6D3A0Ih7u8HmHJB2SdOiVl4518tTLq6l/pE8qXWa2wqwq8dznA++VdAlwIvAa4IvAakmrUm1jI3Ak7X8EOB04LGkVcArwfPNJI2IUGAVYe8bZPf2nuxOHmfW60moiEXFdRGyMiDOBQeA7EfF+4F7gfWm33cBdaXlfWidt/070WfuO+0bMrNe0lUQknd9OWZs+BVwjaZx6n8cNqfwG4NRUfg1wbcHz9yz3jZhZr2m3Oet/AJvbKGspIv4a+Ou0/CTwthb7/D3wL9uMZ8VyE5eZ9ZJ5k4ikrcB5wGmSrslteg0wUGZgZmZWfQvVRI4HTkr7nZwr/zXT/RpmZtan5k0iEfE3wN9I+lpEPLVMMfWVVvNpmZn1inb7RE6QNAqcmT8mIi4sI6i+kp4XqQ24ddDMek+7SeTrwDD16UsmywunPzV3pme1E0nuaDezSmv3OZGJiNgTEQ9FxMPZT6mR9TEP9TWzXtFuEvlzSR+RtF7S2uyn1Mj6nGsgZtYL2m3Oyp4k/0SuLIDXdTYcMzPrJW0lkYg4q+xAzMys97SVRCRd3qo8Im7sbDhmZtZL2m3Oemtu+URgG/A9wEnEzKyPtduc9Uf5dUmrgVvLCMjMzHpH0angfwO4n8TMrM+12yfy5zTewccA8DvA7WUF1c88DYqZ9ZJ2+0T+a255AngqIg6XEE/faiSPNA2KarVGMvGT62ZWVW01Z6WJGH9CfSbfNcBvywyqL+XeuS6pvj56P4MjY35y3cwqq903G14GPET9pVGXAQ9K8lTwHdZc25BqSKW9wdjMbMnabc76NPDWiDgKIOk04K+AO8oKzMzMqq/dP3NrWQJJnl/EsWZmtkK1WxP5tqS/AG5J6zuAu8sJyczMesVC71h/PbAuIj4h6Q+AC9Km+4Gbyg7OzMyqbaGayBeA6wAi4pvANwEk/eO07Z+XGJuZmVXcQv0a6yLih82FqezM+Q6UdKKkhyT9QNKPJP1xKj9L0oOSxiXdJun4VH5CWh9P2+c9v5mZdd9CSWT1PNtetcCxrwAXRsSbgXOAiySdC3wO+HxEvB54Abgy7X8l8EIq/3zaz8zMKmyhJHJI0gebCyV9AJj39bhR91JaPS79BHAh00OD9wKXpuXtaZ20fZv8mLaZWaUt1CfyceBOSe9nOmlsAY4H/sVCJ5c0kI57PfAl4GfAixExkXY5DGxIyxuApwEiYkLSMeBU4Jft/jJmZra85k0iEfEccJ6k3wfelIr/V0R8p52TR8QkcE6aOv5O4A1LiBUASUPAEMA/WLNuqaczM7MlaPd9IvcC9xb9koh4UdK9wFZgtaRVqTayETiSdjsCnA4clrQKOIX6Q43N5xoFRgHWnnG2J5UyM+ui0p46l3RaqoEg6VXAO4HHqSejbN6t3cBdaXlfWidt/0545sEZspl+fVnMrCrKnLpkPXCvpEeB7wL3RMS3gE8B10gap97ncUPa/wbg1FR+DXBtibH1pIhgx/ABJxEzq4x2pz1ZtIh4FHhLi/Ingbe1KP976rME2zw8YM3MqsSTKJqZWWFOImZmVpiTiJmZFeYkYmZmhTmJmJlZYU4iPcDPh5hZVTmJVFxEMDk56edDzKySnESqLoJdIwcRfj7EzKrHSaQH5B8wdNOWmVWJk0iviWDnyEEnETOrBCeRHuSpT8ysKpxEzMysMCeRHuR+ETOrCieRHpEljrTifhEzqwQnkV6RhvqS8ob7RcysCkp7n4h1XnPiyGonkhrbIqJRQ8mXm5mVwTWRHjXXk+wRweDIGIMjY27uMrPSOYn0qnmeZJdqSP5Pa2bl852mh7XTVOWRXGZWJieRFS4iPHmjmZXGHesrRNah3ipZuHPdzMrimsgK4RqHmXWDk8gK4hqHmS230pKIpNMl3Svpx5J+JOljqXytpHskPZE+16RySbpe0rikRyVtLis2MzPrjDJrIhPAv42INwLnAldLeiNwLbA/IjYB+9M6wMXApvQzBOwpMbYVw6OvzKybSksiEfFMRHwvLf8d8DiwAdgO7E277QUuTcvbgRuj7gFgtaT1ZcW3YngeLTPromXpE5F0JvAW4EFgXUQ8kzY9C6xLyxuAp3OHHU5lzecaknRI0qFXXjpWXtA9xH0hZtYtpScRSScB3wA+HhG/zm+L+p/Pi/oTOiJGI2JLRGw54aRTOhhp75sx06+Z2TIoNYlIOo56ArkpIr6Zip/LmqnS59FUfgQ4PXf4xlRmC2gkj6aZfp1UzKxsZY7OEnAD8HhE/Pfcpn3A7rS8G7grV355GqV1LnAs1+xl88klj/xsvpOTkwwOH1hkXc/MrH1lPrF+PvCHwA8lPZLK/j3wWeB2SVcCTwGXpW13A5cA48DLwBUlxrbizOoXySZozCWV5mnjzcyWqrQkEhEHoMUUs3XbWuwfwNVlxdOP8skimyL+tg9d4CRiZh3jJ9b7jJ8pMbNOchLpJ36mxMw6zEmkD+RHabkpy8w6yUmkHzQN/TUz6xQnkT7RPErLTVpm1glOIv3G/SJm1kFOIn3I/SJm1ilOImZmVpiTSB+br3/EfSdm1g4nkT6UJYipqSl2DB9oLOcTht/ZbmbtcBLpR7nOdUlzJgz3nZjZQpxE+lh+mngnDDMrosxZfK3Ksll+a7VGU9bk5OSM2omZ2UKcRPqYpBlTxu8cHqM2MACCmz+4tdvhmVkPcBKxRlNW/V0jtbkn8Dcza+I+EZtT1sSVNXOZmTVzErG5RbBzeIzB4TEnETNryUnE5pW9TncpDx/6wUWzlctJxGbJv38kX1b04UM/uGi2crlj3WbLjdjKW8qzJM3H5hNVrVbzcypmPco1EWup1U29k81SEcGOPfe5v8WsxzmJWPty06XkE0p+Lq7FJISsv8XMeldpSUTSVyUdlfRYrmytpHskPZE+16RySbpe0rikRyVtLisuW5r8GxKzfo6IYHBkjMGRuWsV7lw3W5nKrIl8DbioqexaYH9EbAL2p3WAi4FN6WcI2FNiXLZIzR3t2Xq+FiHV6g8qzn0Sv1HRbAUqLYlExN8Cv2oq3g7sTct7gUtz5TdG3QPAaknry4rNFil1tBPZajA4fICYal27mOshxaUOFTaz6lnuPpF1EfFMWn4WWJeWNwBP5/Y7nMpmkTQk6ZCkQ6+8dKy8SG2GfDNWoxYSweDwGJOTkzO2TU1NsWPPfezYc4CJiYnG9mwfD/c1Wzm61rEe9bvIou8kETEaEVsiYssJJ51SQmQ2r6ZaiWB6vek9JQJ2Do+xc3hs1n/p5udQzKw3LXcSeS5rpkqfR1P5EeD03H4bU5lV0ELPjzS/p2TWCKymRGRmvWu5k8g+YHda3g3clSu/PI3SOhc4lmv2sl7SZoLw0F6zlaG0J9Yl3QK8A3itpMPAfwQ+C9wu6UrgKeCytPvdwCXAOPAycEVZcVn5lpogsmHDfo7ErPpKSyIRsXOOTdta7BvA1WXFYtXQak6uufbbMXyA2z50gZOIWcX5iXVbPovoC3HyMOsNnoDRltVcyWGuWoqbtsyqzTURq4RsQsYdew74uRKzHuIkYpWRPVuya+TgjKfhXQMxqy4nEauc7Gl4z7VlVn3uE7HKytdAmid9dD+JWTW4JmKVlZ+scXJycsbU8+4nMasG10Ssq+Z9diT3mt5dIwdRrdaojbgGYlYNTiLWXRHsGr2fiKmWz49kySLfT3LLVeelQ6cTkJu5zLrDzVnWdQu+0KpJljia36jY3Mzld5eYlc9JxHpL81T0KQG1etti0b4TJx+z9jmJWM/JvyExk71tkZiZBIq8TTEiuGzPfbPezGhmszmJWG9qMQ9XljDyI7nquy4+KQj8nIpZG5xErGc1v7I3rTSSy4yXY1FPClNTU433v2e1k/xrfbNt+fO3w01g1q88Ost6X24oMEyP5No1cpDawMCMXScnJ9k1epBabQAEtwydR0Swc/Qgtwyd12gSk0QQcw4pbp4Y0tPXW79yTcRWhFY37ll9J1myQWk0mJicnGwkjlkJI4LB4bGWzWBzddrnayNZ09pS+lZcw7GqcxKxlW3WaC7N2ibUso8FWjeDZTf0Wc1pufm+Gn0ze+5jcHhsSUnET+dblbk5y1a8+ZqX8tvm2i+r0QwOH0CqcduHLwCmaxoAO0fG6jWc2nTtZ3D4QD1B0aKWk+SbxbL15v3cPGZV5pqI2QLyz6A012R2Do+xc3iskSyyxJJPIM01lKwTP/tpnhOseXv+GNdIrGqcRMwWkmvqyt/QgZaJpdFE1iRrEssSRZY8mmsak5OTXLbnPiYmJur9NVP1qWEGR8ZmjS6bnJxkYmKCiYmJRlk7yaZT/TXZ93fjmZpe7S9aaX8UuDnLrA35zvb55vqasW9e/jjUSEKzJp9snnSyMeKsRhCN0WUgbh7aWl/O+ntqNW656jwGR8a49arzqdVq6ZQz+3Aao8n23IdU49YPnd/4+lqt1nbzWXYOov7d+fNkyTX7rvzNcilzm+Wb/5Y6Iq65KTH7bzHXNWj1qub8wI12j8um6wG49arzO9Jc2c3XSLsmYrZIi53ra9ZxWc1manYNp76fZnw25EeXQW6kmWbcPAQMDo/VazIjYwwOH2BwuF6LyT90me3fagBAc7NbvsaTr3nkvzc7z86RgzNqTRMTE42a1WIGCbSqLTU/OJofgbdQrai55pIftND8euZWNYV8c2MW11yDJ/Kx5697dt5sdGD+WaX871pkloVuDcCoVE1E0kXAF4EB4CsR8dkuh2RWiuZaQVbDaNUM1uq45uVZ+1FPMgOrjiN/yizB3Dy0tV7Q9IxN/iHM93/lASKCmz5w7owaT/ry6XM0nWdWrSk9d5NN5599h6QZf70311iywQkgbrnqvMa1yv8O2bX7Vzc8SExNEVOBatODH/K1hOxGe+tV5ze2zbqeQWOm6J2jB2fVcvKDLGq1ASKmGok8P9ACqNfSUo0xH/Ou0YMMDBzXGEKezUq9c2RsxsCNLNbm72/1srZ8zTZ/PVv+2+hwTaUySUTSAPAl4J3AYeC7kvZFxI+7G5lZyZpu5MVPM3M+sVY3G5hOMLNqPGmgQGZg1XHA1PS+TeE1x9yybyh3XGM6/+w7NJ0cshibE1eWVPNxSZrxO8xIlqr/jvlRc1kSysxIbukdNc2ysizhZfHNHGSRNRdOzfhDoJE4UlJqJNdW1yz3u+UT+YxYv3w/N39wK7u+fH8jqeRrUztHxhqJenB4jFs/VN9nx/CBxnE3f3BrPdFGcPMHc8m/A1SVjh1JW4HPRMS70vp1ABHxn+c6Zs0ZZ8e2T4y0/R21gfpfDjG1+N+56LH+zmoeu2K/M6Yg19Q247imbWXFu6jjYuYNvLbquPqxk5NLizU7b3aO/HrTdVCtNruPK9unKb788S2/M/9988Q/7zVqFWv6bMSaT3xN3yPBTUPn1RPa1PTx+Wv7jY9tezgitswZ4CJUKYm8D7goIj6Q1v8Q+KcR8dGm/YaAobT6JuCxZQ10Ya8FftntIJpUMSaoZlyOqT2OqX1VjOvsiDi5EyeqTHNWuyJiFBgFkHSoU9m0UxxT+6oYl2Nqj2NqXxXjknSoU+eq0uisI8DpufWNqczMzCqqSknku8AmSWdJOh4YBPZ1OSYzM5tHZZqzImJC0keBv6A+xPerEfGjBQ4bLT+yRXNM7atiXI6pPY6pfVWMq2MxVaZj3czMek+VmrPMzKzHOImYmVlhlU4ikn4h6YeSHsmGpElaK+keSU+kzzWpXJKulzQu6VFJmzsYx1clHZX0WK5s0XFI2p32f0LS7hJi+oykI+l6PSLpkty261JMP5X0rlz5RalsXNK1S4zpdEn3SvqxpB9J+lgq79q1miemrl0rSSdKekjSD1JMf5zKz5L0YDr/bWmACZJOSOvjafuZC8XawZi+Junnuet0Tipfln/n6XwDkr4v6VtpvWvXaYG4unqt1KH75aJjyh7lr+IP8AvgtU1lfwJcm5avBT6Xli8B/jf1mQTOBR7sYBxvBzYDjxWNA1gLPJk+16TlNR2O6TPAv2ux7xuBHwAnAGcBP6M+eGEgLb8OOD7t88YlxLQe2JyWTwb+T/rurl2reWLq2rVKv+9Jafk44MH0+98ODKbyYeDDafkjwHBaHgRumy/WDsf0NeB9LfZfln/n6ZzXADcD30rrXbtOC8TV1WtFB+6XRWKqdE1kDtuBvWl5L3BprvzGqHsAWC1pfSe+MCL+FvjVEuN4F3BPRPwqIl4A7gEu6nBMc9kO3BoRr0TEz4Fx4G3pZzwinoyI3wK3pn2LxvRMRHwvLf8d8DiwgS5eq3limkvp1yr9vi+l1ePSTwAXAnek8ubrlF2/O4BtkjRPrJ2MaS7L8u9c0kbg3cBX0rro4nWaK64FLMu1mue7S/1/r+pJJIC/lPSw6tOdAKyLiGfS8rPAurS8AXg6d+xh5r9ZLNVi41iu+D6aqqdfzaqu3YgpNSW8hfpftJW4Vk0xQRevVWoKeQQ4Sv1/1J8BL0bERIvzN747bT8GnFp2TBGRXaf/lK7T5yWd0BxT03d3+r/dF4BPAtlkUafS5es0R1yZbl6rTtwvFx1T1ZPIBRGxGbgYuFrS2/Mbo17/6voY5arEAewB/hFwDvAM8N+6EYSkk4BvAB+PiF/nt3XrWrWIqavXKiImI+Ic6jMzvA14w3J+fyvNMUl6E3Ad9djeSr2J41PLFY+k9wBHI+Lh5frOdswTV9euVdKV+2Wlk0hEHEmfR4E7qf/P9lzWTJU+j6bdl3valMXGUXp8EfFcuhFMAV9musq+bDFJOo76zfqmiPhmKu7qtWoVUxWuVYrjReBeYCv1JoXsAeD8+RvfnbafAjy/DDFdlJoDIyJeAf6U5b1O5wPvlfQL6s2HF1J/31C3r9OsuCT9WZevVaful4uPqVVHSRV+gFcDJ+eWD1Jvm/svzOwo+pO0/G5mdhQ91OF4zmRmJ/ai4qD+l8nPqXdWrUnLazsc0/rc8r+h3g4M8LvM7Fh8knpH8aq0fBbTncW/u4R4BNwIfKGpvGvXap6YunatgNOA1Wn5VcB9wHuArzOzw/gjaflqZnYY3z5frB2OaX3uOn4B+Oxy/ztP530H0x3YXbtOC8TVtWtFh+6XRWJa8kUs64f6KJgfpJ8fAZ9O5acC+4EngL/KfsF0Mb5EvW35h8CWDsZyC/Umj/9HvY3wyiJxAP+aeqfeOHBFCTH9z/Sdj1Kfdyx/o/x0iumnwMW58kuoj1j6WXaNlxDTBdSry48Cj6SfS7p5reaJqWvXCvgnwPfTdz8G/Ifcv/mH0u/8deCEVH5iWh9P21+3UKwdjOk76To9BvwZ0yO4luXfee6c72D6Zt2167RAXF27VnTwfrnYmDztiZmZFVbpPhEzM6s2JxEzMyvMScTMzApzEjEzs8KcRMzMrDAnETMzK8xJxMzMCvv/EJ/2P73fTzIAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.histplot(sl)\n",
    "plt.ylim(top=700)\n",
    "plt.xlim((500,5000))\n",
    "plt.savefig('../misc/plots/plot_protsize.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "db22a059-5fb0-450f-857b-1a862110282d",
   "metadata": {},
   "outputs": [],
   "source": [
    "curated_data=[]\n",
    "for n,node in enumerate(proteome_data):\n",
    "    if len(node['sequence']) < 1500:\n",
    "        curated_data.append(node)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0fa95c2-363d-463f-b629-df81b062e703",
   "metadata": {},
   "source": [
    "## Categorias 'Omniprescentes'\n",
    "\n",
    "\n",
    "La idea es determinar cual de todas las categorías están presentes en todos los datos, para poder luego decidir cuál campo uso como UID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "93953d20-a1e4-4e0b-aafa-2074b90c6f3b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['uniref100',\n",
       " 'taxonomy',\n",
       " 'sequence',\n",
       " 'organism',\n",
       " 'proteome',\n",
       " 'length',\n",
       " 'acc',\n",
       " 'uniparc',\n",
       " 'ncbi_taxon_id']"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_keys=[]\n",
    "for node in curated_data:\n",
    "    all_keys+=list(node.keys())\n",
    "count={}\n",
    "for key in all_keys:\n",
    "    count[key]=0\n",
    "for node in curated_data:\n",
    "    for k in list(node.keys()):\n",
    "        count[k]+=1\n",
    "count_of_omniprescence=len(curated_data)\n",
    "omnipresent_fields=[]\n",
    "for key, value in count.items():\n",
    "    if value == count_of_omniprescence:\n",
    "        omnipresent_fields.append(key)\n",
    "omnipresent_fields"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f95e067f-80cd-46e4-8554-76e4a30592c9",
   "metadata": {},
   "source": [
    "## Determino los identificadores de localización\n",
    "\n",
    "\n",
    "A partir de los datos almacenados en `curated_data` determino cuales datos tienen información sobre la localización subcelular de la proteina. En esta base de datos, cada proteina cuenta con una lista describiendo esta información. Tomo entonces los nodos de esta lista, para tratar de determinar cuales son todas las localizaciones posibles, almacenados en la variable `all_descriptors`.\n",
    "\n",
    "Algunas de estas localizaciones describen conjuntos de localización especifica y general con el formato \"ubicación-general:ubicación-general\", por ejemplo: '\\[80 kda soluble fragment\\]: secreted' o '\\[beta-dystroglycan\\]: cell membrane'. En estos casos me interesa particularmente la ubicación general. Almaceno esta información en las variables `colon_descriptors` y `colon_descriptor_categories`.\n",
    "\n",
    "Luego voy a tomar la información almacenada en `colon_descriptor_categories` para filtrar las proteinas por localización."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "15566bed-d859-4597-a22a-55d92c748e4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "localized=[]\n",
    "unlocalized=0\n",
    "locations=[]\n",
    "for protein in curated_data:\n",
    "    try:\n",
    "        localized.append(protein['localization'])\n",
    "    except KeyError as e:\n",
    "        if e.args[0]=='localization':\n",
    "            unlocalized+=1\n",
    "\n",
    "for location in localized:\n",
    "    locations+=location\n",
    "locations=set(locations)\n",
    "all_descriptors=[]\n",
    "for location in locations:\n",
    "    mezza=[loc.strip().lower() for loc in re.split(r\";|,\", location)]\n",
    "    all_descriptors+=mezza\n",
    "all_descriptors=set(all_descriptors)\n",
    "\n",
    "colon_descriptor_categories=set()\n",
    "colon_descriptors=set()\n",
    "for node in all_descriptors:\n",
    "    if ':' in node:\n",
    "        colon_descriptors.add(node.lower())\n",
    "        parts=node.split(':')\n",
    "        assert len(parts) <= 2\n",
    "        colon_descriptor_categories.add(parts[1].strip().lower())\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "0a233058-ced8-4207-a0ac-a94c6eadb121",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'cell membrane',\n",
       " 'cell surface',\n",
       " 'cytoplasm',\n",
       " 'cytoplasmic vesicle',\n",
       " 'cytoplasmic vesicle membrane',\n",
       " 'endomembrane system',\n",
       " 'endoplasmic reticulum',\n",
       " 'endoplasmic reticulum membrane',\n",
       " 'golgi apparatus',\n",
       " 'golgi apparatus membrane',\n",
       " 'lysosome',\n",
       " 'lysosome lumen',\n",
       " 'membrane',\n",
       " 'membrane raft',\n",
       " 'mitochondrion',\n",
       " 'mitochondrion inner membrane',\n",
       " 'mitochondrion matrix',\n",
       " 'mitochondrion membrane',\n",
       " 'nucleus',\n",
       " 'secreted',\n",
       " 'virion'}"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "colon_descriptor_categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "8ddb0607-f49a-4075-826b-8a51318982a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "surface=[]\n",
    "mitochondrion=[]\n",
    "other=[]\n",
    "cytoplasm=[]\n",
    "for node in sorted(list(colon_descriptor_categories)):\n",
    "    if 'mitoch' in node:\n",
    "        mitochondrion.append(node)\n",
    "    elif 'cell' in node or 'membrane' in node  and 'endo' not in node and 'golgi' not in node and 'cyto' not in node:\n",
    "        surface.append(node)\n",
    "    elif 'golgi' in node or 'lyso' in node or 'endop' in node:\n",
    "            other.append(node)\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "ea9e750e-3956-4914-abad-9a2c63e2df9e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([],\n",
       " ['endoplasmic reticulum',\n",
       "  'endoplasmic reticulum membrane',\n",
       "  'golgi apparatus',\n",
       "  'golgi apparatus membrane',\n",
       "  'lysosome',\n",
       "  'lysosome lumen'],\n",
       " ['cell membrane', 'cell surface', 'membrane', 'membrane raft'],\n",
       " ['mitochondrion',\n",
       "  'mitochondrion inner membrane',\n",
       "  'mitochondrion matrix',\n",
       "  'mitochondrion membrane'])"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytoplasm,other, surface, mitochondrion"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07a187c5-5be7-4285-97f9-c082ba71b3a0",
   "metadata": {},
   "source": [
    "## Formateo los datos \n",
    "\n",
    "Voy a intentar reorganizar los datos para poder facilitar el procesamiento subsecuente. Para ello en un primer lugar genero un diccionario que llamaré `urefed_located_data` que contenga el `uref100` de las proteinas como clave y los datos como valores. \n",
    "\n",
    "Luego voy a eliminar los datos que _a priori_ no considero necesarios generando el diccionario `categorised_data` que contiene el `uref100` de las proteinas como UID nuevamente, y como valores solo voy a conservar localización y secuencia. También voy a agregar un campo `cats` que luego almacenará las categorías en las que se asignará la proteina, con la idea de no tener proteinas almacenadas en dos categorías distintas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "2aeb232f-e87f-498b-9eef-0d9828209695",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "137037 52912\n"
     ]
    }
   ],
   "source": [
    "urefed_curated_data={}\n",
    "cnte=0\n",
    "cnt=0\n",
    "for node in curated_data:\n",
    "    try:\n",
    "        _=node['localization']\n",
    "        urefed_curated_data[node['uniref100']]=node\n",
    "        cnt+=1\n",
    "    except KeyError:\n",
    "        cnte+=1\n",
    "        pass\n",
    "\n",
    "print(cnte,cnt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "e79714e3-de72-464f-8c21-596ba7d1deff",
   "metadata": {},
   "outputs": [],
   "source": [
    "breaks=0\n",
    "categorised_data=[]\n",
    "for key, values in  urefed_curated_data.items():\n",
    "    for value in values['localization']:\n",
    "        for val in re.split(r\",|:|;\", value): \n",
    "            if val.lower() in colon_descriptors or val.lower() in colon_descriptor_categories: \n",
    "                categorised_data.append({\n",
    "                'uniref100':values['uniref100'],\n",
    "                'localization':values['localization'],\n",
    "                'sequence':values['sequence'],\n",
    "                'cats':{'number':0,\n",
    "                       'names':[]}\n",
    "                })\n",
    "            breaks+=1\n",
    "            break\n",
    "        break\n",
    "       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "d266ac9a-b46b-4a7c-9d39-af78b83d9958",
   "metadata": {},
   "outputs": [],
   "source": [
    "labeled_data={\n",
    "    'other':{},\n",
    "    'mitochondrion':{},\n",
    "    'surface':{}\n",
    "}\n",
    "deletable=[]\n",
    "for datum in categorised_data:\n",
    "    for value in datum['localization']:\n",
    "        if value.lower() in other: #['nada']:\n",
    "            labeled_data['other'][datum['uniref100']]=datum\n",
    "            if 'other' not in datum['cats']['names']:\n",
    "                datum['cats']['number']+=1\n",
    "                datum['cats']['names'].append('other')\n",
    "        if value.lower() in mitochondrion:\n",
    "            labeled_data['mitochondrion'][datum['uniref100']]=datum\n",
    "            if 'mitochondrion' not in datum['cats']['names']:\n",
    "                datum['cats']['number']+=1\n",
    "                datum['cats']['names'].append('mitochondrion')\n",
    "        if value.lower() in surface:\n",
    "            labeled_data['surface'][datum['uniref100']]=datum\n",
    "#             datum['cats']['number']+=1\n",
    "            if 'surface' not in datum['cats']['names']:\n",
    "                datum['cats']['number']+=1\n",
    "                datum['cats']['names'].append('surface')\n",
    "        if datum['cats']['number'] > 1:\n",
    "            deletable.append(datum)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "0b77ab1b-0938-4cb4-961c-9f1070f8be4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{('surface', 'other'): 133,\n",
       " ('other', 'surface', 'mitochondrion'): 2,\n",
       " ('other', 'mitochondrion'): 64,\n",
       " ('mitochondrion', 'surface'): 7,\n",
       " ('surface', 'other', 'mitochondrion'): 4,\n",
       " ('surface', 'mitochondrion'): 55,\n",
       " ('other', 'surface'): 84,\n",
       " ('mitochondrion', 'other'): 18,\n",
       " ('surface', 'mitochondrion', 'other'): 4,\n",
       " ('mitochondrion', 'other', 'surface'): 10}"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "local_combinations=set()\n",
    "counts={}\n",
    "for node in deletable:\n",
    "    local_combinations.add(tuple(node['cats']['names']))\n",
    "for node in local_combinations:\n",
    "    counts[node]=0\n",
    "for node in deletable:\n",
    "    counts[tuple(node['cats']['names'])]+=1\n",
    "counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "a21ca10a-f8ec-4395-b961-9100fd6926b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "clean_labeled_data=labeled_data.copy()\n",
    "for repseq in deletable:\n",
    "    cats=repseq['cats']['names']\n",
    "    name=repseq['uniref100']\n",
    "    for cat in cats:\n",
    "        try:\n",
    "            del clean_labeled_data[cat][name]\n",
    "        except KeyError:\n",
    "            pass\n",
    "            #print(cat,name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "9d0e773a-4844-4ccd-8509-bbee708f3dc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('pickle/json_files/fully_preprocessed.json', 'w') as file:\n",
    "    json.dump(clean_labeled_data, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "c1bd7478-3e5a-4ca5-958a-6f1ef9e57df1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clase other tiene 1102 datos\n",
      "clase mitochondrion tiene 1654 datos\n",
      "clase surface tiene 1090 datos\n"
     ]
    }
   ],
   "source": [
    "for k,v in clean_labeled_data.items():\n",
    "    print('clase {} tiene {} datos'.format(k, len(v)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "d13c135f-081b-4a20-b0a1-4087fa840ef1",
   "metadata": {},
   "outputs": [],
   "source": [
    "mitochondrion_proteins=np.array([ protein['sequence'] for uref, protein in clean_labeled_data['mitochondrion'].items() ])\n",
    "surface_proteins=np.array([ protein['sequence'] for uref, protein in clean_labeled_data['surface'].items() ])\n",
    "other_proteins=np.array([ protein['sequence'] for uref, protein in clean_labeled_data['other'].items() ])\n",
    "all_prot_array=np.hstack([surface_proteins,mitochondrion_proteins,other_proteins])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbf6b6da-a84f-4cf4-805f-8293ed2b57f8",
   "metadata": {},
   "source": [
    "### Acomodo clases\n",
    "\n",
    "Como la cantidad de proteinas está desbalanceada en ambas clases, voy a recortar la clase \"mitochondrion\" para que tenga la misma longitud que \"surface\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "id": "445f2cd2-a45d-4714-ac84-516a632af69e",
   "metadata": {},
   "outputs": [],
   "source": [
    "catlength=np.min([len(mitochondrion_proteins), len(surface_proteins), len(other_proteins) ])\n",
    "mitochondrion_proteins=np.random.choice(mitochondrion_proteins,catlength, replace=False)\n",
    "surface_proteins=np.random.choice(surface_proteins,catlength, replace=False)\n",
    "other_proteins=np.random.choice(other_proteins,catlength, replace=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29e91185-6931-4e46-be4d-770b7ed55d75",
   "metadata": {},
   "source": [
    "## Codificación numérica\n",
    "\n",
    "Obtengo todas las posibles variantes de aa en el dataset, le asigno un valor numérico y transformo los arrays en valores numéricos. El código queda guardado en la variable `numeric_code` y en el archivo `numeric_code.json`. Los arrays se llaman `{organela}_proteins_numeric`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "e7be7b1b-5abf-4b72-94a5-7872bf030cde",
   "metadata": {},
   "outputs": [],
   "source": [
    "allseqs=''\n",
    "for sequence in all_prot_array:\n",
    "    allseqs+=sequence\n",
    "aas=set()\n",
    "for val in allseqs:\n",
    "    aas.add(val)\n",
    "numeric_code={aa:num for num,aa in enumerate(aas,3)}\n",
    "with open('pickle/json_files/numeric_code.json', 'w') as file:\n",
    "    json.dump(numeric_code, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "1aa91a63-454a-4e76-aade-8455adb5ce0e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_19416/3002189762.py:7: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray\n",
      "  surface_proteins_numeric=np.array(surface_proteins_numeric)\n",
      "/tmp/ipykernel_19416/3002189762.py:14: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray\n",
      "  mitochondrion_proteins_numeric=np.array(mitochondrion_proteins_numeric)\n"
     ]
    }
   ],
   "source": [
    "surface_proteins_numeric=[]\n",
    "for sequence in surface_proteins:\n",
    "    sequence_array=np.zeros(len(sequence), dtype=int)\n",
    "    for n,amino_acid in enumerate(sequence):\n",
    "        sequence_array[n]=numeric_code[amino_acid]\n",
    "    surface_proteins_numeric.append(sequence_array)\n",
    "surface_proteins_numeric=np.array(surface_proteins_numeric)\n",
    "mitochondrion_proteins_numeric=[]\n",
    "for sequence in mitochondrion_proteins:\n",
    "    sequence_array=np.zeros(len(sequence), dtype=int)\n",
    "    for n,amino_acid in enumerate(sequence):\n",
    "        sequence_array[n]=numeric_code[amino_acid]\n",
    "    mitochondrion_proteins_numeric.append(sequence_array)\n",
    "mitochondrion_proteins_numeric=np.array(mitochondrion_proteins_numeric)\n",
    "other_proteins_numeric=[]\n",
    "for sequence in other_proteins:\n",
    "    sequence_array=np.zeros(len(sequence), dtype=int)\n",
    "    for n,amino_acid in enumerate(sequence):\n",
    "        sequence_array[n]=numeric_code[amino_acid]\n",
    "    other_proteins_numeric.append(sequence_array)\n",
    "other_proteins_numeric=np.array(mitochondrion_proteins_numeric)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f775ca2c-66b9-4b12-9a40-e2dfb5b082ce",
   "metadata": {},
   "source": [
    "## Manejo de arrays final\n",
    "\n",
    "Para tener un dataset uniforme, paddeo las secuencias para que todos los arrays tengan la misma longitud, y hago un vstack para que en lugar de tener un array de arrays (deprecadisimo) tenga un 2D-array. Luego genero los labels acorde a la organela a la que pertenezca la secuencia y guardo los files en la carpeta `src/pickle/numpy_arrays` con formato `.npy` para cada dataset individualmente y todo junto como un paquete `.npz`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "229c7f3e-8959-45b1-b258-c5b406d4c1e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "padded_surface_proteins=np.vstack(pad_sequences(surface_proteins_numeric,padding='post', maxlen=1500))\n",
    "padded_mitochondrion_proteins=np.vstack(pad_sequences(mitochondrion_proteins_numeric,padding='post',maxlen=1500))\n",
    "padded_other_proteins=np.vstack(pad_sequences(other_proteins_numeric,padding='post',maxlen=1500))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "73e5ec1b-de78-475b-9e51-e7e141b904ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((1090, 1500), (1090, 1500), (1090, 1500))"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "padded_other_proteins.shape,padded_surface_proteins.shape, padded_mitochondrion_proteins.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "e6a001f4-00ac-47f6-b524-6c830c1bad3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "data=np.vstack([padded_surface_proteins, padded_mitochondrion_proteins, padded_other_proteins])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "d418b575-e2df-4159-8445-3ed43baa73f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3270, 1500)"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "id": "f6d4b709-d98c-414e-80c0-2e691a460cb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.save('pickle/numpy_arrays/padded_surface_proteins', padded_surface_proteins)\n",
    "np.save('pickle/numpy_arrays/padded_mitochondrion_proteins',padded_mitochondrion_proteins)\n",
    "np.save('pickle/numpy_arrays/padded_other_proteins',padded_other_proteins)\n",
    "np.save('pickle/numpy_arrays/labels', labels)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "id": "18f7b28b-8c01-4c66-884d-c3270d813171",
   "metadata": {},
   "outputs": [],
   "source": [
    "# labels=np.hstack([np.zeros(padded_surface_proteins.shape[0]),np.ones(padded_mitochondrion_proteins.shape[0])]).reshape(-1,1)\n",
    "labels=np.hstack((np.full(catlength, 0),np.full(catlength, 1),np.full(catlength, 2))).reshape(-1,1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "cb75f41b-cbbb-4893-b4b3-2d23ce86cb8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0],\n",
       "       [0],\n",
       "       [0],\n",
       "       ...,\n",
       "       [2],\n",
       "       [2],\n",
       "       [2]])"
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "e9b7f7cf-4a5a-4bfa-be89-040b3a126094",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.savez('pickle/numpy_arrays/preprocessed_data', data=data, labels=labels)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:bio_ml]",
   "language": "python",
   "name": "conda-env-bio_ml-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
